src/aof.c

```diff
@@ -41,189 +41,774 @@
+    if (redis_fsync(fd) == -1) {
+        serverLog(LL_WARNING, "Fail to fsync the temp AOF file %s: %s.", 
+            tmp_am_name, strerror(errno));
+
+        ret = C_ERR;
+        goto cleanup;
+    }
+
+    if (rename(tmp_am_filepath, am_filepath) != 0) {
+        serverLog(LL_WARNING,
+            "Error trying to rename the temporary AOF manifest file %s into %s: %s",
+            tmp_am_name, am_name, strerror(errno));
+
+        ret = C_ERR;
     }
+    
+cleanup:
+    if (fd != -1) close(fd);
+    sdsfree(am_name);
+    sdsfree(am_filepath);
+    sdsfree(tmp_am_name);
+    sdsfree(tmp_am_filepath);
+    return ret;
+}
+
+/* Persist the aofManifest information pointed to by am to disk. */
+int persistAofManifest(aofManifest *am) {
+    if (am->dirty == 0) {
+        return C_OK;
+    }
+
+    sds amstr = getAofManifestAsString(am);
+    int ret = writeAofManifestFile(amstr);
+    sdsfree(amstr);
+    if (ret == C_OK) am->dirty = 0;
+    return ret;
+}
+
+/* Called in `loadAppendOnlyFiles` when we upgrade from a old version redis.
+ * 
+ * 1) Create AOF directory use 'server.aof_dirname' as the name.
+ * 2) Use 'server.aof_filename' to construct a BASE type aofInfo and add it to 
+ *    aofManifest, then persist the manifest file to AOF directory.
+ * 3) Move the old AOF file (server.aof_filename) to AOF directory.
+ *
+ * If any of the above steps fails or crash occurs, this will not cause any 
+ * problems, and redis will retry the upgrade process when it restarts.
+ */
+void aofUpgradePrepare(aofManifest *am) {
+    serverAssert(!aofFileExist(server.aof_filename));
+
+    /* Create AOF directory use 'server.aof_dirname' as the name. */
+    if (dirCreateIfMissing(server.aof_dirname) == -1) {
+        serverLog(LL_WARNING, "Can't open create append-only dir %s", 
+            server.aof_dirname);
+        exit(1);
+    }
+    
+    /* Manually construct a BASE type aofInfo and add it to aofManifest. */
+    if (am->base_aof_info) aofInfoFree(am->base_aof_info);
+    aofInfo *ai = aofInfoCreate();
+    ai->file_name = sdsnew(server.aof_filename);
+    ai->file_seq = 1;
+    ai->file_type = AOF_FILE_TYPE_BASE;
+    am->base_aof_info = ai;
+    am->curr_base_file_seq = 1;
+    am->dirty = 1;
+
+    /* Persist the manifest file to AOF directory. */ 
+    if (persistAofManifest(am) != C_OK) {
+        exit(1);
+    }
+
+    /* Move the old AOF file to AOF directory. */
+    sds aof_filepath = makePath(server.aof_dirname, server.aof_filename);
+    if (rename(server.aof_filename, aof_filepath) == -1) {
+        serverLog(LL_WARNING,
+            "Error trying to move the old AOF file %s into dir %s: %s",
+            server.aof_filename,
+            server.aof_dirname,
+            strerror(errno));
+        sdsfree(aof_filepath);
+        exit(1);;
+    }
+    sdsfree(aof_filepath);
+
+    serverLog(LL_NOTICE, "Redis enters the upgrade mode and successfully completes the upgrade preparation!!");
```

Comment:
```suggestion
    serverLog(LL_NOTICE, "Successfully migrated an old-style AOF file (%s) into the AOF directory (%s).", server.aof_filename, server.aof_dirname);
```

Updated the log message to be more clear to users.

Reply:
OK